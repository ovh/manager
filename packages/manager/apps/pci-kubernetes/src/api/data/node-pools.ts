import { v6 } from '@ovh-ux/manager-core-api';

type TRawClusterNodePool = {
  id: string;
  projectId: string;
  name: string;
  flavor: string;
  status: 'READY' | 'UPDATING' | 'DELETING' | 'INSTALLING' | 'ERROR';
  sizeStatus: string;
  autoscale: boolean;
  monthlyBilled: boolean;
  antiAffinity: boolean;
  desiredNodes: number;
  minNodes: number;
  maxNodes: number;
  currentNodes: number;
  availableNodes: number;
  upToDateNodes: number;
  createdAt: string;
  updatedAt: string;
  autoscaling: {
    scaleDownUtilizationThreshold: number;
    scaleDownUnneededTimeSeconds: number;
    scaleDownUnreadyTimeSeconds: number;
  };
  template: {
    metadata: {
      labels: Record<string, unknown>;
      annotations: Record<string, unknown>;
      finalizers: [];
    };
    spec: {
      unschedulable: boolean;
      taints: [];
    };
  };
};

export type TClusterNodePool = Pick<
  TRawClusterNodePool,
  | 'id'
  | 'name'
  | 'antiAffinity'
  | 'availableNodes'
  | 'desiredNodes'
  | 'autoscale'
  | 'monthlyBilled'
  | 'createdAt'
  | 'status'
  | 'flavor'
  | 'minNodes'
  | 'maxNodes'
> & {
  formattedFlavor: string;
  search?: string;
};

export const getClusterNodePools = async (
  projectId: string,
  clusterId: string,
): Promise<TClusterNodePool[]> => {
  const { data: items } = await v6.get<TRawClusterNodePool[]>(
    `/cloud/project/${projectId}/kube/${clusterId}/nodepool`,
  );

  return items.map(
    (item) =>
      ({
        id: item.id,
        name: item.name,
        antiAffinity: item.antiAffinity,
        availableNodes: item.availableNodes,
        desiredNodes: item.desiredNodes,
        autoscale: item.autoscale,
        monthlyBilled: item.monthlyBilled,
        createdAt: item.createdAt,
        status: item.status,
        flavor: item.flavor,
        minNodes: item.minNodes,
        maxNodes: item.maxNodes,
      } as TClusterNodePool),
  );
};

export const deleteNodePool = async (
  projectId: string,
  clusterId: string,
  poolId: string,
) =>
  v6.delete(`/cloud/project/${projectId}/kube/${clusterId}/nodepool/${poolId}`);

export type TUpdateNodePoolSizeParam = {
  desiredNodes: number;
  minNodes: number;
  maxNodes: number;
  autoscale: boolean;
};

export const updateNodePoolSize = async (
  projectId: string,
  clusterId: string,
  poolId: string,
  param: TUpdateNodePoolSizeParam,
) =>
  v6.put(
    `/cloud/project/${projectId}/kube/${clusterId}/nodepool/${poolId}`,
    param,
  );

export type TCreateNodePoolParam = {
  flavorName: string;
  name: string;
  antiAffinity: boolean;
  monthlyBilled: boolean;
  autoscale: boolean;
  minNodes: number;
  desiredNodes: number;
  maxNodes: number;
};

export const createNodePool = (
  projectId: string,
  clusterId: string,
  param: TCreateNodePoolParam,
) => v6.post(`/cloud/project/${projectId}/kube/${clusterId}/nodepool`, param);
