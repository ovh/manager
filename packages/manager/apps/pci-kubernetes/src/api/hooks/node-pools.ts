import { useMutation, useQuery } from '@tanstack/react-query';
import { PaginationState } from '@ovh-ux/manager-react-components';
import { applyFilters, Filter } from '@ovh-ux/manager-core-api';
import { useMemo, useRef } from 'react';
import { useTranslation } from 'react-i18next';
import { ColumnSort } from '@tanstack/react-table';
import { format, parseISO } from 'date-fns';
import * as dateFnsLocales from 'date-fns/locale';
import { getDateFnsLocale } from '@ovh-ux/manager-core-utils';
import {
  deleteNodePool,
  getClusterNodePools,
  TClusterNodePool,
  TUpdateNodePoolSizeParam,
  updateNodePoolSize,
} from '@/api/data/node-pools';
import { useKubernetesCluster } from '@/api/hooks/useKubernetes';
import { useRegionFlavors } from '@/api/hooks/flavors';
import { compareFunction, paginateResults } from '@/helpers';

export const useClusterNodePools = (projectId: string, clusterId: string) => {
  const { i18n } = useTranslation('common');
  const locales = useRef({ ...dateFnsLocales }).current;
  const userLocale = getDateFnsLocale(i18n.language);
  return useQuery({
    queryKey: ['project', projectId, 'kubernetes', clusterId, 'nodePools'],
    queryFn: async () => {
      const pools = await getClusterNodePools(projectId, clusterId);
      return pools.map((pool) => {
        const createdAt = parseISO(pool.createdAt);
        return {
          ...pool,
          createdAt: format(createdAt, 'dd MMM yyyy HH:mm:ss', {
            locale: locales[userLocale],
          }),
        };
      });
    },
    throwOnError: true,
  });
};

export const sortClusterNodePools = (
  pools: TClusterNodePool[],
  sorting: ColumnSort,
): TClusterNodePool[] => {
  const data = [...pools];

  if (sorting) {
    const { id: sortKey, desc } = sorting;

    data.sort(
      compareFunction<TClusterNodePool>(sortKey as keyof TClusterNodePool),
    );
    if (desc) {
      data.reverse();
    }
  }

  return data;
};

export const usePaginatedClusterNodePools = (
  projectId: string,
  clusterId: string,
  pagination: PaginationState,
  filters: Filter[],
  sorting: ColumnSort,
) => {
  const { t } = useTranslation('detail');
  const {
    data: pools,
    error: poolsError,
    isPending: isPoolsPending,
    isLoading: isPoolsLoading,
  } = useClusterNodePools(projectId, clusterId);

  const {
    data: cluster,
    error: clusterError,
    isPending: isClusterPending,
    isLoading: isClusterLoading,
  } = useKubernetesCluster(projectId, clusterId);

  const {
    data: flavors,
    error: flavorsError,
    isPending: isFlavorsPending,
    isLoading: isFlavorsLoading,
  } = useRegionFlavors(projectId, cluster?.region);

  const poolsWithFlavors = useMemo(
    () =>
      (pools || []).map((pool) => {
        const flavor = (flavors || []).find((f) => f.name === pool.flavor);

        const formattedFlavor = t('kube_flavor', {
          name: flavor?.name?.toUpperCase(),
          cpuNumber: flavor?.vcpus,
          ramCapacity: flavor?.ram / 1000,
          diskCapacity: flavor?.disk,
        });

        return {
          ...pool,
          formattedFlavor,
          search: `${pool.name} ${formattedFlavor}`,
        };
      }),
    [pools, flavors, t],
  );

  return {
    data: useMemo(
      () =>
        paginateResults<TClusterNodePool>(
          applyFilters(
            sortClusterNodePools(poolsWithFlavors || [], sorting),
            filters,
          ),
          pagination,
        ),
      [poolsWithFlavors, pagination, filters, sorting],
    ),
    error: poolsError || clusterError || flavorsError,
    isLoading: isPoolsLoading || isClusterLoading || isFlavorsLoading,
    isPending: isPoolsPending || isClusterPending || isFlavorsPending,
  };
};

type ActionNodePoolProps = {
  projectId: string;
  clusterId: string;
  poolId: string;
  onError: (cause: Error) => void | Promise<void>;
  onSuccess: () => void | Promise<void>;
};

export const useDeleteNodePool = ({
  projectId,
  clusterId,
  poolId,
  onError,
  onSuccess,
}: ActionNodePoolProps) => {
  const mutation = useMutation({
    mutationFn: async () => deleteNodePool(projectId, clusterId, poolId),
    onError: (cause: Error) => {
      onError(cause);
    },
    onSuccess: async () => {
      onSuccess();
    },
  });

  return {
    deletePool: () => mutation.mutate(),
    ...mutation,
  };
};

export const useUpdateNodePoolSize = ({
  projectId,
  clusterId,
  poolId,
  onError,
  onSuccess,
}: ActionNodePoolProps) => {
  const mutation = useMutation({
    mutationFn: async (param: TUpdateNodePoolSizeParam) =>
      updateNodePoolSize(projectId, clusterId, poolId, param),
    onError: (cause: Error) => {
      onError(cause);
    },
    onSuccess: async () => {
      onSuccess();
    },
  });

  return {
    updateSize: (param: TUpdateNodePoolSizeParam) => mutation.mutate(param),
    ...mutation,
  };
};
