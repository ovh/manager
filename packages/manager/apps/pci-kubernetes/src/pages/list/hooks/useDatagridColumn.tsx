import { useHref, useParams } from 'react-router-dom';

import { useTranslation } from 'react-i18next';

import { ODS_THEME_COLOR_INTENT } from '@ovhcloud/ods-common-theming';
import { OsdsLink } from '@ovhcloud/ods-components/react';

import { Clipboard, DataGridTextCell, DatagridColumn } from '@ovh-ux/manager-react-components';

import ActionsComponent from '@/components/listing/actions.component';
import ClusterStatus from '@/components/service/ClusterStatus.component';
import use3AZPlanAvailable from '@/hooks/use3azPlanAvaible';
import { TKube } from '@/types';

import Mode from '../components/Mode.component';

export const useDatagridColumn = () => {
  const { t } = useTranslation(['listing', 'kube']);
  const { projectId = '' } = useParams();
  const has3AZ = use3AZPlanAvailable();

  const columns: DatagridColumn<TKube>[] = [
    {
      id: 'name',
      cell: (props: TKube) => (
        <DataGridTextCell>
          <OsdsLink color={ODS_THEME_COLOR_INTENT.primary} href={useHref(`${props.id}`)}>
            {props.name}
          </OsdsLink>
        </DataGridTextCell>
      ),
      label: t('kube_list_name'),
    },
    {
      id: 'id',
      cell: (props: TKube) => (
        <DataGridTextCell>
          <Clipboard aria-label="clipboard" value={props.id} />
        </DataGridTextCell>
      ),
      label: t('kube_list_id'),
    },
    {
      id: 'plan',
      cell: (props: TKube) => (
        <DataGridTextCell>{t(`kube:kube_service_cluster_plan_${props.plan}`)}</DataGridTextCell>
      ),
      label: t('kube:kube_service_cluster_plan'),
    },
    {
      id: 'region',
      cell: (props: TKube) => <DataGridTextCell>{props.region}</DataGridTextCell>,
      label: t('kube_list_region'),
    },
    ...(has3AZ
      ? [
          {
            id: 'mode',
            cell: (props: TKube) => <Mode projectId={projectId} region={props.region} />,
            label: t('kubernetes_containers_deployment_mode_label'),
          },
        ]
      : []),
    {
      id: 'attachedTo',
      cell: (props: TKube) => <DataGridTextCell>{props.attachedTo}</DataGridTextCell>,
      label: t('kube_list_network_attached'),
    },
    {
      id: 'version',
      cell: (props: TKube) => <DataGridTextCell>{props.version}</DataGridTextCell>,
      label: t('kube_list_version'),
    },
    {
      id: 'status',
      cell: (props: TKube) => (
        <DataGridTextCell>
          <ClusterStatus status={props.status} />
        </DataGridTextCell>
      ),
      label: t('kube:kube_service_cluster_status'),
    },
    {
      id: 'actions',
      cell: (props: TKube) => (
        <div className="min-w-16">
          <ActionsComponent kubeId={props.id} />
        </div>
      ),
      label: '',
    },
  ];

  return columns;
};
