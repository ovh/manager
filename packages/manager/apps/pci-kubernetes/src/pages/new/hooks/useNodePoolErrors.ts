import { useEffect, useState } from 'react';

import { ensureNameIsUnique } from '@/helpers';
import { isNodePoolNameValid } from '@/helpers/matchers/matchers';
import { NodePoolPrice, NodePoolState } from '@/types';

function generateUniqueNameWithZone(
  baseName: string,
  zoneName: string | null,
  existingNodePools: NodePoolPrice[],
  multipleNodes: boolean,
) {
  const zoneSuffix = zoneName?.split('-').pop();
  const nameWithSuffix = zoneSuffix && multipleNodes ? `${baseName}-${zoneSuffix}` : baseName;
  ensureNameIsUnique(nameWithSuffix, existingNodePools);
}

function validateNodePoolNameUniqueness({
  nodePoolName,
  selectedAvailabilityZones,
  nodes,
}: {
  nodePoolName: string;
  selectedAvailabilityZones?: { zone: string; checked: boolean }[] | null;
  nodes: NodePoolPrice[];
}) {
  if (!selectedAvailabilityZones?.length) {
    ensureNameIsUnique(nodePoolName, nodes);
    return;
  }

  const filterZones = selectedAvailabilityZones
    .filter((zone) => zone.checked)
    .map(({ zone }) => zone);

  if (!filterZones.length) {
    ensureNameIsUnique(nodePoolName, nodes);
    return;
  }

  filterZones.forEach((zoneName: string) => {
    generateUniqueNameWithZone(nodePoolName, zoneName, nodes, filterZones.length > 1);
  });
}

export function useNodePoolErrors(nodePoolState: NodePoolState, nodes: NodePoolPrice[] | null) {
  const [error, setError] = useState<{ valid?: string | null; exists?: string | null } | null>(
    null,
  );

  const isValidName = isNodePoolNameValid(nodePoolState.name);

  useEffect(() => {
    if (!nodes?.length) {
      return setError((prevError) => ({ ...prevError, exists: null }));
    }
    try {
      validateNodePoolNameUniqueness({
        nodePoolName: nodePoolState.name,
        selectedAvailabilityZones: nodePoolState.selectedAvailabilityZones,
        nodes,
      });
      setError((prevError) => ({ ...prevError, exists: null }));
    } catch (err) {
      setError((prevError) => ({
        ...prevError,
        exists: 'kube_add_node_pool_name_already_exist_validation_error',
      }));
    }
  }, [nodePoolState.selectedAvailabilityZones, nodePoolState.name, nodes]);

  useEffect(() => {
    setError((prevError) => {
      if (!isValidName && nodePoolState.isTouched) {
        return { ...prevError, valid: 'kube_add_node_pool_name_input_pattern_validation_error' };
      }
      return { ...prevError, valid: null };
    });
  }, [isValidName, nodePoolState.isTouched]);

  const handleCreationError = (err: unknown) => {
    if (err instanceof Error && err.message === 'name already exists') {
      setError((prevError) => ({
        ...prevError,
        exists: 'kube_add_node_pool_name_already_exist_validation_error',
      }));
    }
  };

  const clearExistsError = () => {
    setError((prevError) => ({
      ...prevError,
      exists: null,
    }));
  };

  return {
    error,
    isValidName,
    handleCreationError,
    clearExistsError,
  };
}
