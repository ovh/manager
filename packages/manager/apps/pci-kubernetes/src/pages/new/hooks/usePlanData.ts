import { useCallback } from 'react';

import { useCatalog } from '@ovh-ux/manager-pci-common';

import { DeploymentMode, TClusterPlan, TClusterPlanEnum } from '@/types';

type TPlan = {
  title: string;
  type: DeploymentMode;
  description: string;
  content: string[];
  footer?: string;
  value: TClusterPlan;
  code: string;
  price: number | null;
};

const usePlanData = (): { plans: TPlan[]; isPending: boolean } => {
  const { data: catalog, isPending: isPendingCatalog } = useCatalog();

  const pricing = useCallback(
    (code: string) => {
      if (catalog) {
        const getAddon = catalog.addons.find((add) => add.planCode === code);

        return getAddon?.pricings[0].price ?? null;
      }
      return null;
    },
    [catalog],
  );

  const plans: TPlan[] = [
    {
      title: 'kube_add_plan_title_free',
      description: 'kube_add_plan_description_free',
      content: [
        'kube_add_plan_content_free_control',
        'kube_add_plan_content_free_high_availability',
        'kube_add_plan_content_free_SLO',
        'kube_add_plan_content_free_auto_scaling',
        'kube_add_plan_content_free_ETCD',
        'kube_add_plan_content_free_version',
        'kube_add_plan_content_free_100',
      ],
      value: TClusterPlanEnum.FREE,
      type: DeploymentMode.MONO_ZONE,
      code: 'mks.free.hour.consumption',
    },
    {
      title: 'kube_add_plan_title_standard',
      description: 'kube_add_plan_description_standard',
      content: [
        'kube_add_plan_content_standard_3AZ_control_plane',
        'kube_add_plan_content_standard_disponibility',
        'kube_add_plan_content_standard_SLA',
        'kube_add_plan_content_free_auto_scaling',
        'kube_add_plan_content_standard_ETCD',
        'kube_add_plan_content_standard_version',
        'kube_add_plan_content_standard_500',
      ],
      type: DeploymentMode.MULTI_ZONES,
      value: TClusterPlanEnum.STANDARD,
      code: 'mks.standard.hour.consumption.3az',
    },
  ].map((plan) => ({
    ...plan,
    price: pricing(plan.code),
  }));

  return {
    isPending: isPendingCatalog,
    plans,
  };
};

export default usePlanData;
