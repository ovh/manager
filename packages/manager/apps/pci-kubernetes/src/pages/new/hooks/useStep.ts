import { useState } from 'react';

export interface StepState {
  isOpen: boolean;
  isChecked: boolean;
  isLocked: boolean;
}

export function useStep(initialState?: Readonly<Partial<StepState>>) {
  const [isOpen, setIsOpen] = useState(!!initialState?.isOpen);
  const [isChecked, setIsChecked] = useState(!!initialState?.isChecked);
  const [isLocked, setIsLocked] = useState(!!initialState?.isLocked);
  return {
    isOpen,
    isChecked,
    isLocked,
    open: () => setIsOpen(true),
    close: () => setIsOpen(false),
    check: () => setIsChecked(true),
    uncheck: () => setIsChecked(false),
    lock: () => setIsLocked(true),
    unlock: () => setIsLocked(false),
  };
}
