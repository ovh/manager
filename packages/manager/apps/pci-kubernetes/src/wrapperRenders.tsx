import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import {
  ShellContext,
  ShellContextType,
} from '@ovh-ux/manager-react-shell-client';
import { vi } from 'vitest';

export const shellContext = {
  environment: {
    getRegion: vi.fn().mockReturnValue('EU'),
    getUser: () => ({
      ovhSubsidiary: 'mocked_ovhSubsidiary',
      currency: { code: 'EUR' },
    }),
  },
  shell: {
    navigation: {
      getURL: vi.fn(),
    },
  },
};

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
});

export const wrapper = ({ children }) => (
  <QueryClientProvider client={queryClient}>
    <ShellContext.Provider
      value={(shellContext as unknown) as ShellContextType}
    >
      {children}
    </ShellContext.Provider>
  </QueryClientProvider>
);
