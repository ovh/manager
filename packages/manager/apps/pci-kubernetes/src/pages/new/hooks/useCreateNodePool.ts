import { useCallback, useEffect, useState } from 'react';

import { useParam } from '@ovh-ux/manager-pci-common';
import { convertHourlyPriceToMonthly } from '@ovh-ux/manager-react-components';

import { useRegionInformations } from '@/api/hooks/useRegionInformations';
import { TComputedKubeFlavor } from '@/components/flavor-selector/FlavorSelector.component';
import { NODE_RANGE, TAGS_BLOB } from '@/constants';
import { checkIfNameExists, generateUniqueName } from '@/helpers';
import { isNodePoolNameValid } from '@/helpers/matchers/matchers';
import { hasInvalidScalingOrAntiAffinityConfig } from '@/helpers/node-pool';
import useMergedFlavorById, { getPriceByDesiredScale } from '@/hooks/useMergedFlavorById';
import { NodePoolPrice, NodePoolState } from '@/types';

function generateUniqueNameWithZone(
  baseName: string,
  zoneName: string | null,
  existingNodePools: NodePoolPrice[],
  multipleNodes: boolean,
) {
  const zoneSuffix = zoneName?.split('-').pop();
  const nameWithSuffix = zoneSuffix && multipleNodes ? `${baseName}-${zoneSuffix}` : baseName;

  const nameExists = checkIfNameExists(nameWithSuffix, existingNodePools);

  if (nameExists) {
    throw new Error('Le nom de nodepool est déjà utilisé. Veuillez en choisir un autre.');
  }
  return generateUniqueName(nameWithSuffix, existingNodePools);
}

const useCreateNodePools = ({ name, isLocked }: { name?: string; isLocked: boolean }) => {
  const [nodes, setNodes] = useState<NodePoolPrice[] | null>(null);
  const [selectedFlavor, setSelectedFlavor] = useState<TComputedKubeFlavor | null>(null);

  const [isMonthlyBilled, setIsMonthlyBilled] = useState(false);
  const [error, setError] = useState<string | null>(null);
  const [nodePoolState, setNodePoolState] = useState<NodePoolState>({
    antiAffinity: false,
    name: '',
    isTouched: false,
    scaling: {
      quantity: { desired: NODE_RANGE.MIN, min: 0, max: NODE_RANGE.MAX },
      isAutoscale: false,
    },
  });
  const { projectId } = useParam('projectId');
  const [nodePoolEnabled, setNodePoolEnabled] = useState(true);
  const { data: regionInformations } = useRegionInformations(projectId, name ?? null);
  const selectedAvailabilityZones = nodePoolState.selectedAvailabilityZones?.filter(
    ({ checked }) => checked,
  );

  const price = useMergedFlavorById<{ hour: number; month?: number } | null>(
    projectId,
    name ?? null,
    selectedFlavor?.id ?? null,
    {
      select: (flavor) =>
        getPriceByDesiredScale(
          flavor.pricingsHourly?.price,
          flavor.pricingsMonthly?.price,

          nodePoolState.scaling?.quantity.desired * (selectedAvailabilityZones?.length ?? 1),
        ),
    },
  );

  const createNodePool = useCallback(() => {
    if (!nodes || !nodePoolState.scaling || !selectedFlavor) return;

    const selectedZones = nodePoolState.selectedAvailabilityZones?.filter((zone) => zone.checked);

    const createDataNodePool = ({ zone }: { zone?: string } = {}): NodePoolPrice => ({
      name: generateUniqueNameWithZone(
        nodePoolState.name,
        zone ?? null,
        nodes,
        !!(selectedZones?.length && selectedZones?.length >= 2),
      ),
      antiAffinity: nodePoolState.antiAffinity,
      autoscale: nodePoolState.scaling.isAutoscale,
      ...(zone && { availabilityZones: [zone] }),
      localisation: zone ?? name ?? null,
      desiredNodes: nodePoolState.scaling.quantity.desired,
      ...(nodePoolState.scaling.isAutoscale && {
        minNodes: nodePoolState.scaling.quantity.min,
        maxNodes: nodePoolState.scaling.quantity.max,
      }),
      flavorName: selectedFlavor.name ?? '',
      monthlyPrice: isMonthlyBilled
        ? (price?.month ?? 0)
        : convertHourlyPriceToMonthly(price?.hour ?? 0),
      monthlyBilled: isMonthlyBilled,
    });

    try {
      const newNodePools = selectedZones?.length
        ? selectedZones.filter((zone) => zone.checked).map(createDataNodePool)
        : [createDataNodePool()];

      setNodes([...nodes, ...newNodePools]);

      setNodePoolState((state) => ({
        ...state,
        name: '',
        isTouched: false,
      }));
    } catch (err: unknown) {
      if (err instanceof Error) {
        setError(err.message);
      }
    }
  }, [nodes, nodePoolState, selectedFlavor, name, isMonthlyBilled, price]);

  const isValidName = isNodePoolNameValid(nodePoolState.name);

  const isNodePoolValid = !nodePoolEnabled || (Boolean(selectedFlavor) && isValidName);

  const isButtonDisabled =
    !isNodePoolValid ||
    (regionInformations &&
      hasInvalidScalingOrAntiAffinityConfig(regionInformations?.type, nodePoolState));

  const isPricingComingSoon = selectedFlavor?.blobs?.tags?.includes(TAGS_BLOB.COMING_SOON);

  const isStepUnlocked = !isLocked;

  const canSubmit =
    (isStepUnlocked && !nodePoolEnabled) ||
    (isStepUnlocked && nodePoolEnabled && Array.isArray(nodes) && nodes.length > 0);

  useEffect(() => setIsMonthlyBilled(false), [selectedFlavor]);

  useEffect(
    () =>
      setError(
        !isValidName && nodePoolState.isTouched
          ? 'kube_add_node_pool_name_input_pattern_validation_error'
          : null,
      ),
    [isValidName, nodePoolState.isTouched],
  );

  useEffect(() => {
    if (regionInformations?.availabilityZones.length) {
      setNodePoolState((state) => ({
        ...state,
        selectedAvailabilityZones: regionInformations?.availabilityZones.map((zone) => ({
          zone,
          checked: true,
        })),
      }));
    }
  }, [regionInformations?.availabilityZones, setNodePoolState]);

  useEffect(() => {
    setNodes(!nodePoolEnabled ? null : []);
    if (!nodePoolEnabled) {
      setNodePoolState((state) => ({
        ...state,

        name: '',
        isTouched: false,
      }));
    }
  }, [nodePoolEnabled, setNodePoolState, setNodes]);

  const onDelete = useCallback(
    (nameToDelete: string) => nodes && setNodes(nodes.filter((node) => node.name !== nameToDelete)),
    [nodes],
  );

  return {
    state: {
      nodes,
      nodePoolState,
      selectedFlavor,
      isMonthlyBilled,
      nodePoolEnabled,
      price,
    },

    actions: {
      setNodes,
      setNodePoolState,
      setSelectedFlavor,
      setIsMonthlyBilled,
      setNodePoolEnabled,
      createNodePool,
      onDelete,
    },

    view: {
      isValidName,
      error,
      isNodePoolValid,
      isButtonDisabled,
      isPricingComingSoon,
      isStepUnlocked,
      canSubmit,
    },
  };
};

export default useCreateNodePools;
