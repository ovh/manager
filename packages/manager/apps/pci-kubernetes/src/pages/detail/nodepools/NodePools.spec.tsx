import { vi, describe } from 'vitest';
import { render } from '@testing-library/react';
import * as nodePoolsModule from '@/api/hooks/node-pools';
import * as queryClientModule from '@/queryClient';
import { wrapper } from '@/wrapperRenders';
import Page from './NodePools.page';
import { TClusterNodePool } from '@/api/data/node-pools';

const nodePool: TClusterNodePool = {
  id: '1',
  name: 'NodePool1',
  antiAffinity: false,
  availableNodes: 1,
  desiredNodes: 1,
  autoscale: false,
  monthlyBilled: false,
  createdAt: '2021-10-10T10:10:10',
  status: 'READY',
  flavor: 'flavor',
  minNodes: 1,
  maxNodes: 1,
  formattedFlavor: 'formattedFlavor',
};

describe('NodePools', () => {
  it('should show spinner if nodepools are loading', async () => {
    vi.spyOn(nodePoolsModule, 'usePaginatedClusterNodePools').mockReturnValue({
      isPending: true,
      data: undefined,
      isLoading: false,
      error: undefined,
    });

    const { container } = render(<Page />, {
      wrapper,
    });

    expect(container).toMatchSnapshot();
  });

  it('should render datagrid when nodepools are loaded', async () => {
    vi.spyOn(nodePoolsModule, 'usePaginatedClusterNodePools').mockReturnValue({
      isPending: false,
      data: {
        rows: [nodePool],
        totalRows: 1,
        pageCount: 1,
      },
      isLoading: false,
      error: undefined,
    });

    const { getByText } = render(<Page />, {
      wrapper,
    });

    expect(getByText('NodePool1')).toBeInTheDocument();
  });

  it('should reload nodepools on refresh', () => {
    vi.mock('@/queryClient', () => ({
      default: {
        invalidateQueries: vi.fn(),
      },
    }));

    vi.spyOn(nodePoolsModule, 'usePaginatedClusterNodePools').mockReturnValue({
      isPending: false,
      data: {
        rows: [nodePool],
        totalRows: 1,
        pageCount: 1,
      },
      isLoading: false,
      error: undefined,
    });

    const { getByTestId } = render(<Page />, {
      wrapper,
    });

    const refreshButton = getByTestId('refresh-button');

    refreshButton.click();

    expect(queryClientModule.default.invalidateQueries).toHaveBeenCalledWith({
      queryKey: ['project', 'project-id', 'kubernetes', 'kube-id', 'nodePools'],
    });
  });
});
