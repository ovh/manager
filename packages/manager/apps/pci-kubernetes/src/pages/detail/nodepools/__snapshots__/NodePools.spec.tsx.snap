// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NodePools > should show spinner if nodepools are loading 1`] = `
<div>
  <div
    class="sm:flex items-center justify-between mt-4"
  >
    <div
      class="flex flex-row"
    >
      <osds-button
        class="xs:mb-0.5 sm:mb-0"
        color="primary"
        size="sm"
        variant="stroked"
      >
        kube_node_pool_add
      </osds-button>
      <osds-button
        class="xs:mb-0.5 sm:mb-0 ml-0.5"
        color="primary"
        data-testid="refresh-button"
        size="sm"
        variant="stroked"
      >
        <osds-icon
          class="mr-2"
          color="primary"
          name="refresh"
          size="xs"
        />
      </osds-button>
    </div>
    <div
      class="justify-between flex"
    >
      <osds-search-bar
        class="w-[70%]"
        value=""
      />
      <osds-popover>
        <osds-button
          color="primary"
          size="sm"
          slot="popover-trigger"
          variant="stroked"
        >
          <osds-icon
            class="mr-2"
            color="primary"
            name="filter"
            size="xs"
          />
          common_criteria_adder_filter_label
        </osds-button>
        <osds-popover-content>
          <osds-form-field>
            <div
              slot="label"
            >
              <osds-text
                color="text"
                level="heading"
              >
                common_criteria_adder_column_label
              </osds-text>
            </div>
            <osds-select
              data-testid="add-filter_select_idColumn"
              value="name"
            >
              <osds-select-option
                value="name"
              >
                kube_node_pool_name
              </osds-select-option>
              <osds-select-option
                value="flavor"
              >
                kube_nodes_flavor
              </osds-select-option>
              <osds-select-option
                value="autoscale"
              >
                Autoscaling
              </osds-select-option>
              <osds-select-option
                value="monthlyBilled"
              >
                kube_node_pool_monthly_billing
              </osds-select-option>
              <osds-select-option
                value="createdAt"
              >
                kube_node_pool_creation_date
              </osds-select-option>
            </osds-select>
          </osds-form-field>
          <osds-form-field
            class="mt-2"
          >
            <div
              slot="label"
            >
              <osds-text
                color="text"
                level="heading"
              >
                common_criteria_adder_operator_label
              </osds-text>
            </div>
            <osds-select
              value="includes"
            >
              <osds-select-option
                value="includes"
              >
                common_criteria_adder_operator_includes
              </osds-select-option>
              <osds-select-option
                value="starts_with"
              >
                common_criteria_adder_operator_starts_with
              </osds-select-option>
              <osds-select-option
                value="ends_with"
              >
                common_criteria_adder_operator_ends_with
              </osds-select-option>
              <osds-select-option
                value="is_equal"
              >
                common_criteria_adder_operator_is_equal
              </osds-select-option>
              <osds-select-option
                value="is_different"
              >
                common_criteria_adder_operator_is_different
              </osds-select-option>
            </osds-select>
          </osds-form-field>
          <osds-form-field
            class="mt-2"
          >
            <div
              slot="label"
            >
              <osds-text
                color="text"
                level="heading"
              >
                common_criteria_adder_value_label
              </osds-text>
            </div>
            <osds-input
              class="border"
              color="primary"
              data-testid="filter-add_value-input"
              type="text"
              value=""
            />
          </osds-form-field>
          <osds-button
            class="mt-4"
            color="primary"
            data-testid="filter-add_submit"
            disabled="true"
            size="sm"
          >
            common_criteria_adder_submit_label
          </osds-button>
        </osds-popover-content>
      </osds-popover>
    </div>
  </div>
  <div
    class="my-5"
  />
  <div
    class=""
    data-testid="wrapper"
  >
    <osds-spinner
      data-testid="spinner"
      inline="true"
      size="md"
    />
  </div>
  Outlet
</div>
`;
